src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java

```diff
@@ -0,0 +1,488 @@
+     * on the same cache generation, otherwise the results are discarded. It then goes and fills the relevant data for
+     * the shard (response + failures), issueing a reroute at the end of it to make sure there will be another round
+     * of allocations taking this new data into account.
+     */
+    protected synchronized void processAsyncFetch(ShardId shardId, T[] responses, FailedNodeException[] failures) {
+        if (closed) {
+            // we are closed, no need to process this async fetch at all
+            return;
+        }
+        Map<String, NodeEntry<T>> shardCache = cache.get(shardId);
+        if (shardCache == null) {
+            // node longer with us, nothing to do
+            return;
+        }
+        if (responses != null) {
+            for (T response : responses) {
+                NodeEntry<T> nodeEntry = shardCache.get(response.getNode().getId());
+                // if the entry is there, and not marked as failed already, process it
+                if (nodeEntry != null && nodeEntry.isFailed() == false) {
+                    nodeEntry.doneFetching(response);
+                }
+            }
+        }
+        if (failures != null) {
+            for (FailedNodeException failure : failures) {
+                NodeEntry<T> nodeEntry = shardCache.get(failure.nodeId());
+                // if the entry is there, and not marked as failed already, process it
+                if (nodeEntry != null && nodeEntry.isFailed() == false) {
+                    Throwable unwrappedCause = ExceptionsHelper.unwrapCause(failure.getCause());
+                    // if the request got rejected or timed out, we need to try it again next time...
+                    if (unwrappedCause instanceof EsRejectedExecutionException || unwrappedCause instanceof ReceiveTimeoutTransportException || unwrappedCause instanceof ElasticsearchTimeoutException) {
+                        nodeEntry.restartFetching();
+                    } else {
+                        logger.warn("{}: failed to list shard for {} on node [{}]", failure, shardId, getClass().getSimpleName(), failure.nodeId());
+                        nodeEntry.doneFetching(failure.getCause());
+                    }
+                }
+            }
+        }
+        reroute(shardId, "post_response");
+    }
+
+    /**
+     * Process complete failure for the given nodes ids.
+     */
+    protected final void processAsyncFetch(ShardId shardId, String[] nodesIds, Throwable failure) {
+        FailedNodeException[] failures = new FailedNodeException[nodesIds.length];
+        for (int i = 0; i < failures.length; i++) {
+            failures[i] = new FailedNodeException(nodesIds[i], "total failure in fetching", failure);
+        }
+        processAsyncFetch(shardId, null, failures);
+    }
+
+    /**
+     * Schedules a reroute on the cluster update thread.
+     */
+    protected void reroute(ShardId shardId, String reason) {
```

Comment:
this method seems to be wrong here the class is concerned with fetching data in an async fashion, I think what we do with it or what we call as a result should just be somewhere else. I wonder if we can just have an ActionListener that we pass to execute this onSuccess and onFailure since those seem to be the two mode we have.